BoxControl: Prevent invalid style values

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/33318

## Description

This PR fixes invalid values being saved via the `BoxControl` for block supports such as padding and margin.

#### Changes made in this PR include:

1. Change `BoxControl` to manage unit selections via internal state so they can be separated from the block attributes
2. Filtering values in `onChange` callback to not save invalid styles meant updating how mixed values etc were determined

#### Key aspects of this PR

- Simply filtering out CSS unit only values in the `onChange` handlers broke the existing method of determining mixed values in the box control. 
- Filtering of CSS unit only values meant the unlinked controls had no means of maintaining the user's unit selection unless the attribute for that field had been set
- Majority of the changes are around maintaining the existing behaviour while adding internal state to manage unit selections as a fallback if no attribute value is present.

#### Steps to replicate original issue
1. Add group block to a post, select it, then find the padding control in the sidebar
2. Change the unit on the box control without entering any value
3. Inspect the block or switch to the code editor view and see the CSS unit only values saved in the attributes and styles

## How has this been tested?
Via written tests and manual testing in block editor and [Gutenberg Storybook](http://localhost:50240/?path=/story/components-boxcontrol--default)

`npm run test-unit:watch packages/components/src/box-control/test/`

#### Test instructions

1. Checkout PR and run above tests
2. Add group block to a post, select it, then find the padding control in the sidebar
3. Change the unit on the box control without entering any value
4. Inspect the block or switch to the code editor view and confirm no invalid style values in attributes or inline styles
5. Unlink the BoxControl, adjust a value, save and confirm the block support styles still work

## Screenshots <!-- if applicable -->

| Before | After |
|---|---|
| ![BoxControlBefore](https://user-images.githubusercontent.com/60436221/125736218-716d3e14-58bc-451d-8013-fa3ed35721d4.gif) | ![BoxControlAfter](https://user-images.githubusercontent.com/60436221/125881060-c5e5cc16-0f51-44d9-85f7-42658f9ac24f.gif) |

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Prevent CSS unit only values

This also maintains CSS unit selections between the "all" control and the unlinked controls.

andrewserong commented on packages/components/src/box-control/test/index.js:
```diff
@@ -192,4 +192,117 @@ describe( 'BoxControl', () => {
+			} );
+		} );
+
+		it( 'should clear side attribute when control is cleared', () => {
```
In this test case, the value is cleared by committing it via the ENTER key. Should we add another one for clearing the field, and then switching focus (tabbing or clicking out of it?)

aaronrobertshaw replied:
Thanks. I only added the test for committing the change via the `ENTER` key because I was leaving the behaviour for the field to revert to the original value on blur. 

If its agreed we should change that as well, I'll definitely add a test covering it. 👍

andrewserong replied:
Gotcha, that sounds good to me.

andrewserong replied:
This test is currently failing. Once https://github.com/WordPress/gutenberg/pull/33485 lands, we can then update the BoxControl to set `required={ false }` on its UnitControl components, which should then get this passing again, if I'm following correctly?

aaronrobertshaw replied:
I've removed that test in https://github.com/WordPress/gutenberg/pull/33444/commits/ca2fe1cfbc232da221bd079755c75d9d4aa6b74b.

If the current proposal in https://github.com/WordPress/gutenberg/pull/33485 is accepted we can add the test back along with the `required={ false }` prop, if needed. If the `NumberControl` ends up being changed so that it isn't required by default I'm not sure we need the test.

andrewserong reviewed (commented):
Thanks for working on this @aaronrobertshaw and for rolling out the fix to the axial control, too (and for all the helpful inline comments).

I noticed that #33319 is resolved when I clear the value from the unlinked side and then commit it via the ENTER key, but not if I tab or click out of the field:

<!--
![clicking-out-of-box-control](https://user-images.githubusercontent.com/14988353/125746593-620bdf9b-4314-45e2-8f9d-190f5b63c337.gif)
-->

<img width="400" src="https://user-images.githubusercontent.com/14988353/125746593-620bdf9b-4314-45e2-8f9d-190f5b63c337.gif" />

Other than that, these changes fix the issues with empty units appearing in markup for me, and the other testing steps worked well 👍. And I like the choice to use state for tracking the selected units so we can support the empty values.

Happy to do more testing tomorrow!

aaronrobertshaw commented:
Your reviews @andrewserong are always great and much appreciated 👍 

> I noticed that #33319 is resolved when I clear the value from the unlinked side and then commit it via the ENTER key, but not if I tab or click out of the field:

I did consider this. The existing behaviour was that the original value was retained after a blur event such as tabbing or clicking out of the field. If the field was cleared then committed via the `ENTER` key press, the previous behaviour ended up converting that to a `0` value.

It felt that the existing behaviour on blur could be considered an abandonment of editing the field. Submitting an empty value and having that result in a `0` value set is wrong. In an effort to minimise the impact of these changes, only correcting the behaviour when submitting the empty value on `ENTER` key press looked like the way to go. 

I'm more than happy to alter this to set the empty value on blur as well if others agree. 

There is also the `isPressEnterToChange` prop that gets passed through all the control components. If it is set to true, the blur event will commit changes.

> And I like the choice to use state for tracking the selected units so we can support the empty values.

I couldn't see many other options to be able to maintain the current UX. The downside was this did introduce a range of edge cases resulting in a much larger PR.

andrewserong commented:
> In an effort to minimise the impact of these changes, only correcting the behaviour when submitting the empty value on ENTER key press looked like the way to go.

Thanks @aaronrobertshaw, I agree with your thinking there — better to fix the issue that's clearly not working correctly than make too many assumptions and add too much extra complexity. In my testing, I actually didn't mind the blur behaviour retaining the value, and needing to hit ENTER to commit it.

> There is also the isPressEnterToChange prop that gets passed through all the control components

Ah, yes, thanks for the reminder — good to know it'll be a fairly straightforward change if we do want to add it in. Because the existing behaviour is that blur doesn't commit the value, I don't think this discussion should block this PR from moving forward, we can always address it in a follow-up 🙂

aaronrobertshaw commented:
> Because the existing behaviour is that blur doesn't commit the value, I don't think this discussion should block this PR from moving forward, we can always address it in a follow-up

Thanks for working through the issue with me @andrewserong. 

After some further discussion and thinking, I'll split the changes to the NumberControl out into a separate PR so this one is solely around the BoxControl, its unit selection management and not saving invalid styles. 

The new PR can continue the discussion on the NumberControl and seek some fresh ideas or opinions. I don't believe it's required by the BoxControl changes. I'll link it back here for posterity. After that's done we should be close to merging this.

aaronrobertshaw commented:
I've split out the `NumberControl` changes from this PR into https://github.com/WordPress/gutenberg/pull/33485. While doing that I've also rebased this one.

aaronrobertshaw committed:
Remove test for emptying control

This will be covered by the new test in the NumberControl PR fixing that bug. Once both PRs land a new one can be created updating the use of the BoxControl's unit and number controls adding required tests then.

aaronrobertshaw commented:
Now that https://github.com/WordPress/gutenberg/pull/33485 has been merged, I've rebased this to pick up those changes.

- Programmatic tests all pass
- BoxControl component behaves as expected within the Gutenberg Storybook
- The BoxControl does not produce invalid "CSS unit only" style values

It's worth noting the updates for the `NumberControl` made those fields optional by default. This means the BoxControl now gets the ability to clear values on pressing `ENTER` for free. No need to reintroduce earlier changes or tests on that side of things.

@andrewserong any chance you could give this a final sanity check and approval? While it is a bit of an edge case I'd still like to land this fix before it causes anyone further issues.

andrewserong reviewed (approved):
Thanks for the ping @aaronrobertshaw, this is working nicely for me after the rebase, no stray   "CSS unit only" values, and the Enter key behaviour to commit empty values that we get for free, feels natural to me 👍
